keps/sig-node/4569-cgroup-v1-deprecation/README.md

```diff
@@ -0,0 +1,677 @@
+			return
+		}
+	}
+```
+Revised Code Snippet:
+```golang
+if m.memorySwapBehavior == kubelettypes.LimitedSwap {
+    if !isCgroup2UnifiedMode() && utilfeature.DefaultFeatureGate.Enabled(features.DeprecatedCgroupV1) {
+      klog.Warning("cgroup v1 support has been deprecated, please plan for the migration towards cgroup v2")
+      swapConfigurationHelper.ConfigureNoSwap(lcr)
+      return
+    }
+}
+```
+
+Furthermore, code segments that default to cgroup v1 logic will be inverted to assume cgroup v2 as the default. This shift underscores the transition to cgroup v2 and ensures warning logs are consistently generated for cgroup v1 usage. e.g.
+
+Original Code Snippet:
+```golang
+	memLimitFile := "memory.limit_in_bytes"
+	if libcontainercgroups.IsCgroup2UnifiedMode() {
+		memLimitFile = "memory.max"
+	}
+```
+Revised Code Snippet:
+```golang
+	memLimitFile = "memory.max"
+	if libcontainercgroups.!IsCgroup2UnifiedMode() && utilfeature.DefaultFeatureGate.Enabled(features.DeprecatedCgroupV1) {
+      klog.Warning("cgroup v1 support has been deprecated, please plan for the migration towards cgroup v2")
+      memLimitFile := "memory.limit_in_bytes"
+	}
+```
+
+### Feature Flag and Kubelet Option
+
+During the `alpha` stage of this feature's lifecycle, it's anticipated that users will require time to plan and execute their migrations to cgroup v2. Factors influencing the migration timeline include, but are not limited to, selecting compatible kernels and operating systems, planning for potential downtimes, and coordinating changes to minimize disruption during peak operational periods.
+
+To accommodate this, we propose introducing a kubelet flag, `--suppress-cgroupv1-deprecation-warning`, which is not set by default. This flag provides users the flexibility to suppress repetitive deprecation warning logs related to cgroup v1 while they are in the process of planning and executing their migration to cgroup v2. The introduction of this flag is intended to reduce log spam and operational noise, acknowledging the non-trivial nature of the migration for many users.
+
+#### Beta Stage Behavior Modification
+
+As this feature progresses to beta, the default behavior of the kubelet will evolve to reflect the urgency of the migration. Specifically, the kubelet will, by default, refuse to start on hosts still utilizing cgroup v1, thereby directly encouraging users to migrate to cgroup v2 to continue leveraging Kubernetes' full capabilities.
+
+To balance this stricter enforcement with the need for operational flexibility, the `--suppress-cgroupv1-deprecation-warning` flag introduced during the `alpha` stage will retain its functionality but will be enhanced to allow the kubelet to start on cgroup v1 hosts. Users who require additional time for migration, due to logistical or technical constraints, can explicitly set this flag as a temporary measure to bypass the startup restriction while they complete their transition to cgroup v2.
```

Comment:
It seems to be that maybe we should have two flags.

--fail-cgroup-v1, --suppress-cgroupv1-deprecation-warning?

Reply:
in `Beta` the flag `--suppress-cgroupv1-deprecation-warning` can be interpreted as "ignore the cgroup v1 warning and start the kubelet anyway". 

I would like to keep this as simple as possible. More flags means more things to remember for the user, which might lead to bad UX.

Reply:
Yea, but having ``--suppress-cgroupv1-deprecation-warning`` to really mean that we are going to fail to start on kubelet is also not great UX.

Reply:
We should think of a better name for that flag then. 2 new flags just for the deprecation is little overkill, no?

Reply:
`--allow-cgroupv1-operations`?

Reply:
Yea, that seems better to me.

Reply:
how about `--override-cgroupv1-deprecation`?

Reply:
tbh from my perspective I don't think we need a flag to set a failure. If we eventually plan to fail on cgroupv1, we could use the feature gate as a gate on whether to fail or not, and then independently use the flag to toggle whether we print a warning

Reply:
Yeah, I agree.

Reply:
I can also see the argument that maybe we shouldn't let people turn off deprecation notices.

Reply:
log spam could lead to higher resource utilization.